File: webelexis-server/src/main/java/ch/rgw/elexis/Elexis.java
/*
 * This file is part of Webelexis. Copyright (c) 2015 by G. Weirich.
 */

package ch.rgw.elexis;

import ch.rgw.tools.StringTool;
import ch.rgw.vertx.Util;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.jdbc.JDBCClient;
import io.vertx.ext.sql.SQLConnection;

import java.io.IOException;

/**
 * Created by gerry on 07.07.15.
 */
public class Elexis extends AbstractVerticle {
  private JsonObject cfg;
  private JDBCClient client;
  private SQLConnection conn;
  private boolean bConnected;
  private Elexis theInstance;

  private Elexis() {
    this.cfg = config();
    client = JDBCClient.createShared(vertx, cfg.getJsonObject("sql"));
    client.getConnection(cres -> {
      if (cres.succeeded()) {
        bConnected = true;
        conn = cres.result();
      } else {
        throw new RuntimeException("could not connect to elexis server", cres.cause());
      }

    });
  }

  public Elexis getInstance() throws IOException {
    if (theInstance == null) {
      try {
        theInstance = new Elexis();
      } catch (Throwable t) {
        throw new IOException("failed to create connection ", t.getCause());
      }
    }
    return theInstance;
  }

  public JsonObject getPatient(String id) {
    return null;
  }

  public void addPatient(String name, String firstName, String dob) {
    conn.setAutoCommit(false, acRes -> {
      if (acRes.succeeded()) {
        String pid = StringTool.unique("webelexis");
        String sql = "INSERT INTO KONTAKT (id,Bezeichnung1,Bezeichnung2,Geburtsdatum, istPatient) VALUES(?,?,?,?)";
        conn.updateWithParams(sql, Util.asJsonArray(new String[]{name, firstName, dob, "1"}), pres -> {
          if (pres.succeeded()) {
            String sql2 = "INSERT INTO FAELLE (id";

          } else {
            conn.rollback(rbRes -> {
              conn.setAutoCommit(true, acres -> {
              });
            });
          }
        });
      }
    });
  }

}

File: webelexis-server/src/main/java/ch/rgw/elexis/Elexis.kt
/*
 * This file is part of Webelexis. Copyright (c) 2015 by G. Weirich.
 */

package ch.rgw.elexis

import ch.rgw.tools.JdbcLink
import ch.rgw.tools.StringTool
import ch.rgw.tools.TimeTool
import io.vertx.core.AbstractVerticle
import io.vertx.core.json.JsonObject
import java.io.IOException
import java.sql.SQLException
import java.util.Date

/**
 * Created by gerry on 07.07.15.
 */
public class Elexis private constructor() : AbstractVerticle() {
  private val cfg: JsonObject
  private var theInstance: Elexis? = null
  private val j: JdbcLink

  init {
    this.cfg = config().getJsonObject("sql")
    j = JdbcLink(cfg.getString("driver_class", "com.mysql.jdbc.Driver"), cfg.getString("url", "jdbc:mysql:localhost"), "default")
    if (!j.connect(cfg.getString("user", ""), cfg.getString("password", ""))) {

      throw RuntimeException("could not connect to elexis server: " + j.lastErrorString)
    }
  }

  throws(IOException::class)
  public fun getInstance(): Elexis {
    if (theInstance == null) {
      try {
        theInstance = Elexis()
      } catch (t: Throwable) {
        throw IOException("failed to create connection ", t.getCause())
      }

    }
    return theInstance!!
  }

  public fun getPatient(id: String): JsonObject? {
    return null
  }

  throws(SQLException::class)
  public fun addPatient(name: String, firstName: String, dob: String) {
    val sqlKontakt = "INSERT INTO KONTAKT (id,Bezeichnung1,Bezeichnung2,Geburtsdatum, istPatient,deleted,lastupdate) VALUES(?,?,?,?,?,?,?)"
    val patId = StringTool.unique("webelexis")
    val dobShort = TimeTool(dob).toString(TimeTool.DATE_COMPACT)
    val pK = createPreparedStatement(j, sqlKontakt, patId, name, firstName, dobShort, "0", "1");
    pK.setLong(7, Date().getTime())
    val res = pK.executeUpdate()
    if (res == 1) {
      val fallID = StringTool.unique("webelexis")
      val sqlFaelle = "INSERT INTO FAELLE (id, PatientID, DatumVon, deleted, lastupdate) VALUES(?,?,?,?,?)"
      val pF = createPreparedStatement(j, sqlFaelle, fallID, patId, TimeTool().toString(TimeTool.DATE_COMPACT), "0")
      pF.setLong(5, Date().getTime())
      val r2 = pF.executeUpdate()
      if (r2 == 1) {
        val consID = StringTool.unique("webelexis")
        val sqlKons = "INSERT INTO BEHANDLUNGEN (id,FallID,Datum,deleted,lastupdate) VALUES (?,?,?,?,?)"
        val pC = createPreparedStatement(j, sqlKons, consID, fallID, TimeTool().toString(TimeTool.DATE_COMPACT), "0")
        pC.setLong(5, Date().getTime())
        val r3 = pC.executeUpdate();
      }
    }
  }

}
